### ISSUE TITLE
Stable column order for generated SQL statements

### ISSUE THREAD

**Issue by lafriks** [go-rel/rel#280]
Currently looks like because of map usage for storing column data each time insert/update/select SQL statements are generated column order is different, that makes grouping of SQL statements for perf stats in logs impossible and also this makes it unusable for prepared statement caching being polluted and unusable that hurts performance.

**Comment by Fs02** [comment]
sounds important and need to be fixed

do you have suggestion what to use other than map? maybe sorted slice? 🤔

**Comment by lafriks** [comment]
Something like https://github.com/elliotchance/orderedmap can be used instead of map

**Comment by lafriks** [comment]
At least with this library elements can be looped in order keys were added to map with code:
```go
for el := m.Front(); el != nil; el = el.Next() {
    fmt.Println(el.Key, el.Value)
}
```

**Comment by Fs02** [comment]
although this may increase allocation due to internal linked list, this seems better and easier solution than using sorted slice 🤔 

(feel free to work on this if you have time 🙏 )

**Comment by lafriks** [comment]
Looked a bit into this but problem is that it affects a lot of public API where maps are used to pass around mutations that are of type `map[string]Mutate` so these would need to be changed to custom new type with needed methods that could internally use either ordered map or maybe even array (`[]Mutate`) would be sufficient. Anyway it would be breaking change... At least for manual mutation usage and for adapter public API

**Comment by lafriks** [comment]
problem is how would you initialize them them when currently everywhere map[string]Mutate is used in API?

**Comment by Fs02** [comment]
hmmm true, looks like this mostly used in unit testing in this repo, not very sure about use case outside this repo 🤔 

how about, for now we can do lazy initialization during first `First()` calls, like this?
```
func (m *Mutation) First() *Mutate {
  // check if list is not yet initialized, or the content of map is updated
  if (m.size != len(m.Mutates)) {
   // initialize the list
  }

  return m.first
}
``` 

in the future, I think we can deprecate direct access to the underlying map, and remove it completely in future version
(actually I always wanted to revisit what API should be exported and what's not, I think this is one of case that should be un-exported)

**Comment by Fs02** [comment]
ah forgot that we pass map of Mutates instead of Mutation to adapter, seems breaking change is un-avoidable 🤦 
https://github.com/go-rel/rel/blob/1cac74f8a195c3e0453f62734c3fd780ef2b4228/adapter.go#L15-L17

**Comment by lafriks** [comment]
yes that was my proposal to change `map[string]Mutate` to `Mutates` in interfaces and expose only needed functionality with type functions and hide underlying storage structure

**Comment by lafriks** [comment]
Also why do you avoid using pointers in interfaces? (`Query` vs `*Query`). Passing it by value does impact allocation count imho

**Comment by Fs02** [comment]
> yes that was my proposal to change map[string]Mutate to Mutates in interfaces and expose only needed functionality with type functions and hide underlying storage structure

How will Mutates looks like?
I was thinking to replace it by rel.Mutation actually, seems more flexible in the long run 🤔 

> Also why do you avoid using pointers in interfaces? (Query vs *Query). Passing it by value does impact allocation count imho

as far as I know, unlike C++, in golang it's not about whether passing reference vs passing by value to avoid copying. it's about whether it'll cause stack allocation vs heap allocation (stack is generally the fastest because heap allocation increase load to GC)

and as far as I understand, compiler decide where to allocate using escape analysis, and one of the cause is related to the use of pointer (indirection). so my rule of thumb is basically use plain object as much as possible, use pointer where necessary or it's proven to be better by benchmark.

also created a quick benchmark here (shows calling a function wrapped on interface with ptr cause allocation): https://github.com/Fs02/go-pattern-benchmark/tree/master/pass_value_vs_ptr

**Comment by lafriks** [comment]
I was thinking something like:

```go
func NewMutates([]Mutate) Mutates

type Mutates struct {
  ...
}

func (m *Mutates) Add(mutate Mutate)

func (m *Mutates) List() []Mutate
```

probably other functions would be needed (`Get(name)`, `Remove(name)` etc) but that still needs to be decided in process of refactoring

**Comment by lafriks** [comment]
Only it's that it does duplicate values passed around as fields from `Mutation` are first preprocessed and split to multiple fields that are passed to adapter functions, that would make it a bit hard to use imho

**Comment by Fs02** [comment]
> Yes Mutation can also be reused for this. Just field Mutates needs to be made private and different type not map

if not using map, how will you handle access using field name?

> Only it's that it does duplicate values passed around as fields from Mutation are first preprocessed and split to multiple fields that are passed to adapter functions, that would make it a bit hard to use imho

sorry, I don't quite understand the case

**Comment by Fs02** [comment]
> Yes Mutation can also be reused for this. Just field Mutates needs to be made private and different type not map

if not using map, how will you handle access using field name?

> Only it's that it does duplicate values passed around as fields from Mutation are first preprocessed and split to multiple fields that are passed to adapter functions, that would make it a bit hard to use imho

sorry, I don't quite understand the case

**Comment by lafriks** [comment]
> if not using map, how will you handle access using field name?

I don't seem to be finding anywhere where currently there would be need to access Mutate by field name (only use case currently seems to be only to not allow duplicate mutates for same field).

If that's really needed we could have two struct fields:
```
mutates []Mutate
fields map[string]int

**Comment by Fs02** [comment]
> I don't seem to be finding anywhere where currently there would be need to access Mutate by field name (only use case currently seems to be only to not allow duplicate mutates for same field).

Insert all query use it for example:
https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/go-rel+Mutates%5B&patternType=literal

> If that's really needed we could have two struct fields:
> ```
> mutates []Mutate
> fields map[string]int
> ```

that won't be scalable in the future if we need to have function to delete `Mutate` (in worst case we have to update all fields map value)

I prefer to use solution like orderedmap (map that refer to linkedlist items), which I described here: https://github.com/go-rel/rel/issues/280#issuecomment-1126224150

**Comment by lafriks** [comment]
Even if delete is needed I don't think it's much of a problem:

```go
i, ok := fields[fieldName]
if !ok {
    return
}
mutates = append(mutates[:i], mutates[i+1:]...)
```

**Comment by Fs02** [comment]
basically need to shift the whole array?
I don't see why that's better than map of linked list, which just changing the pointer 🤔

**Comment by lafriks** [comment]
Because it can have errors later on. As mostly in library uses non pointer values to pass around you could have copied Mutate value somewhere that would point to different next that in the original map where it could have been changed (like same deleted element).

**Comment by Fs02** [comment]
> copied Mutate value somewhere that would point to different next that in the original map where it could have been changed (like same deleted element).

hmm, even if we delete the Next value, copied Mutate will still point to a valid pointer, because we can only delete element entry from map (not deallocating the actual element)


> Even if delete is needed I don't think it's much of a problem:
>
> ```i, ok := fields[fieldName]
> if !ok {
>     return
> }
> delete(fields, fieldName)
> mutates = append(mutates[:i], mutates[i+1:]...)

this still have `in worst case we have to update all fields map value`, because after shifting the array, now fields map will have wrong index

but I guess this shouldn't be a problem, better way to delete a`Mutate` is just by moving last element to deleted element

**Comment by Fs02** [comment]
let's go with the `slice` + `map[string]int` solution 🚀 

(seems map of linked list item is not a very clean solution, a lot of pointers involved https://go.dev/play/p/tIgI5M2fWq0)
